### PR TITLE
Add `Context` abstraction

### PR DESCRIPTION
**Pull request by Morriar**
Basically what `TestHelpers::Project` was but with a more generic angle.

A context represents a directory (that can be the root directory of a project).

On a context, one can:
* manipulate files
* run commands (bundle, sorbet, arbitrary command)
* deal with project specific stuffs (Gemfile, sorbet/config, etc.)
* ... the scope can be extended as needed

Other projects such as Tapioca or RBI-Central can re-use Context to implement their own logic.

I only migrated the tests to get a look and feel of the thing.
To migrate the different commands to use a context it will take a bit of refactor I'll do in another PR.

### CODE REVIEW

**Review by KaanOzkan — APPROVED** [review]

File: lib/spoom/context.rb [hunk 1]
```diff
@@ -0,0 +1,215 @@
+# typed: strict
+# frozen_string_literal: true
+
+require "fileutils"
+require "open3"
+
+module Spoom
+  # An abstraction to a Ruby project context
```
Comment (rafaelfranca) [inline lib/spoom/context.rb:8]:
```suggestion
  # An abstraction to a Ruby project context.
```

File: lib/spoom/context.rb [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+  #
+  # A context maps to a directory in the file system.
+  # It is used to manipulate files and run commands in the context of this directory.
+  class Context
+    extend T::Sig
+
+    # The absolute path to the directory this context is about
```
Comment (rafaelfranca) [inline lib/spoom/context.rb:15]:
```suggestion
    # The absolute path to the directory this context is about.
```

File: lib/spoom/context.rb [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+    sig { returns(String) }
+    attr_reader :absolute_path
+
+    # Create a new context in the system's temporary directory
```
Comment (rafaelfranca) [inline lib/spoom/context.rb:19]:
```suggestion
    # Create a new context in the system's temporary directory.
```

File: lib/spoom/context.rb [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+    #
+    # `name` is used as prefix to the temporary directory name.
+    # The directory will be created if it doesn't exist.
+    sig { params(name: T.nilable(String)).returns(T.attached_class) }
+    def self.mktmp!(name = nil)
+      new(::Dir.mktmpdir(name))
+    end
+
+    # Create a new context about `absolute_path`
```
Comment (rafaelfranca) [inline lib/spoom/context.rb:28]:
```suggestion
    # Create a new context about `absolute_path`.
```

File: lib/spoom/context.rb [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+    #
+    # The directory will not be created if it doesn't exist.
+    # Call `#make!` to create it.
+    sig { params(absolute_path: String).void }
+    def initialize(absolute_path)
+      @absolute_path = T.let(::File.expand_path(absolute_path), String)
+    end
+
+    # Returns the absolute path to `relative_path` in the context's directory
```
Comment (rafaelfranca) [inline lib/spoom/context.rb:37]:
```suggestion
    # Returns the absolute path to `relative_path` in the context's directory.
```

File: lib/spoom/context.rb [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+    sig { params(relative_path: String).returns(String) }
+    def absolute_path_to(relative_path)
+      File.join(@absolute_path, relative_path)
+    end
+
+    # File System
+
+    # Create the context directory at `absolute_path`
```
Comment (rafaelfranca) [inline lib/spoom/context.rb:45]:
```suggestion
    # Create the context directory at `absolute_path`.
```

File: lib/spoom/context.rb [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+    sig { void }
+    def mkdir!
+      FileUtils.rm_rf(@absolute_path)
+      FileUtils.mkdir_p(@absolute_path)
+    end
+
+    # List all files in this context matching `pattern`
```
Comment (rafaelfranca) [inline lib/spoom/context.rb:52]:
```suggestion
    # List all files in this context matching `pattern`.
```

File: lib/spoom/context.rb [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+    sig { params(pattern: String).returns(T::Array[String]) }
+    def glob(pattern = "**/*")
+      Dir.glob(absolute_path_to(pattern)).map do |path|
+        Pathname.new(path).relative_path_from(@absolute_path).to_s
+      end.sort
+    end
+
+    # List all files at the top level of this context directory
```
Comment (rafaelfranca) [inline lib/spoom/context.rb:60]:
```suggestion
    # List all files at the top level of this context directory.
```

File: lib/spoom/context.rb [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+      glob("*")
+    end
+
+    # Does `relative_path` point to an existing file in this context directory?
+    sig { params(relative_path: String).returns(T::Boolean) }
+    def file?(relative_path)
+      File.file?(absolute_path_to(relative_path))
+    end
+
+    # Return the contents of the file at `relative_path` in this context directory
```
Comment (rafaelfranca) [inline lib/spoom/context.rb:72]:
```suggestion
    # Return the contents of the file at `relative_path` in this context directory.
```

**Review by rafaelfranca — APPROVED** [review]

File: lib/spoom/context.rb (hunk 1 above)
Comment (Morriar) [reply lib/spoom/context.rb:8]: I think by now most of Spoom is documented without a `.`. I'll do a pass on the documentation at some point.

**Pull request merged**